File: py/sort-colors.py
'''
Given an array with n objects colored red, white or blue, sort them in-place
so that objects of the same color are adjacent, with the colors in the order
red, white and blue.

Here, we will use the integers 0, 1, and 2 to represent the color red, white,
and blue respectively.

Note: You are not suppose to use the library's sort function for this problem.

Example:
Input: [2,0,2,1,1,0]
Output: [0,0,1,1,2,2]
'''


# Two pass solution using an array to count the occurences on the first pass.
# The second pass will reconstruct the new array we wish to return.

# Time Complexity: 2n passes implies this algorithm runs in O(n)
# Space Complexity: O(k), where k is the # of colors we store in color_dict
class Solution0:
    def sortColors(self, nums: [int]) -> None:
        """ Do not return anything, modify nums in-place instead. """
        c0 = c1 = c2 = 0
        for num in nums:
            if num == 0:
                c0 += 1
            elif num == 1:
                c1 += 1
            else:
                c2 += 1
        nums[:c0] = [0] * c0
        nums[c0:(c0 + c1)] = [1] * c1
        nums[c0+c1:] = [2] * c2



# Single pass solution using a left and right pointer

# Time Complexity: a single pass implies this algorithm runs in O(n)
# Space Complexity: O(k), where k is the constant # of colors we must store
class Solution:
    def sortColors(self, nums: [int]) -> None:
        left = curr = 0
        right = len(nums) - 1

        while curr <= right:
            if nums[curr] == 0:
                # Swap left and curr indexes
                nums[left], nums[curr] = nums[curr], nums[left]

                # Advance pointers
                left += 1
                curr += 1
            elif nums[curr] == 2:
                # Swap right with curr
                nums[curr], nums[right] = nums[right], nums[curr]

                # Decrease right hand side pointer
                right -= 1
            else: # If nums[curr] == 1
                curr += 1

nums = [0, 1, 2, 2, 1, 1, 2, 2, 0, 0, 0, 0, 2, 1]
nums = [0]

print("Before Sort: ")
print(nums)
# [0, 1, 2, 2, 1, 1, 2, 2, 0, 0, 0, 0, 2, 1]

Solution().sortColors(nums)
print("After Sort: ")
print(nums)
# [0, 0, 0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 2]

File: cpp/sort-colors.cpp
// We can assume that after we filled in all 0's and 1's,
// the remaining elements up to nums.end() will be filled in with the value 2.

class Solution {
public:
    void sortColors(vector<int>& nums) {
        int c0 = 0;
        int c1 = 0;
        // We dont need this line below (it is commented out)
        // int c2 = 0;

        // Scan every number and count each occurrence of 0 or 1
        for (int num : nums) {
            switch (num) {
            case 0:
                ++c0;
                break;
            case 1:
                ++c1;
                break;
            // case 2:
            //     ++c2;
            //     break;
            }
        }

        // modify nums accordingly
        fill(nums.begin(), nums.begin() + c0, 0);
        fill(nums.begin() + c0, nums.begin() + c0 + c1, 1);
        fill(nums.begin() + c0 + c1, nums.end(), 2);
    }
};


// If we *assume* that the array is evenly distributed (has the same amount of 0's, 1's and 2's) we can see that when:
// i) n=6, there are k=2 2's in the vector, since 6/3 = 2 ( 3 is the amount of total colors )

// Above, we avoided 4 operations for each loop:
//   1. Checking the case where num is 2
//   2. Incrementing the integer
//   3. Assigning it
//   4. Break command to exit the switch statement

// which gives us 2*4=8 operations when n=6. Oh, and the initial variable declaration is 4 bytes and an extra assigment, so we end up with 8+1=9 total operations. 

// This may not seem like much at first, but if we consider a much larger value:
// ii) n=6000, there are k=2000 2's in the vector.

// In this case, we saved 2000*4+1 = 8001 total operations by avoiding extra work.

class Solution {
public:
    void sortColors(vector<int>& nums) {
        int c0 = 0;
        int c1 = 0;
        // We dont need this line below (it is commented out)
        // int c2 = 0;

        // Scan every number and count each occurrence of 0 or 1
        for (int num : nums) {
            switch (num) {
            case 0:
                ++c0;
                break;
            case 1:
                ++c1;
                break;
            // case 2:
            //     ++c2;
            //     break;
            }
        }

        // modify nums accordingly
        fill(nums.begin(), nums.begin() + c0, 0);
        fill(nums.begin() + c0, nums.begin() + c0 + c1, 1);
        fill(nums.begin() + c0 + c1, nums.end(), 2);
    }
};


// One pass solution, use high (right side) and low (left side) pointers.
// Invariant: leave 1's in the middle, swap any 0's to the left and swap any
//            1's to the left.
class Solution {
public:
    void sortColors(vector<int>& nums) {
        int left = 0;
        int curr = 0;
        int right = nums.size() - 1;

        // If curr == right, then we are done.
        while (curr <= right) {
            switch (nums[curr]) {
            case 0:
                // swap left with curr (using std::swap())
                swap(nums[left], nums[curr]);
                ++left;
                ++curr;
                break;
            case 2:
                // swap right with curr
                swap(nums[right], nums[curr]);
                --right;
                break;
            default:
                // advance if nums[curr] == 1
                ++curr;
            }
        }
    }
};



// Space efficient solution using a fixed-size array
class Solution {
public:
    int arr[3];
    int curr = 0;

    void color_array(int color, vector<int>& nums) {
        while (arr[color]--) {
            nums[curr++] = color;
        }
    }

    void sortColors(vector<int>& nums) {
        for (int num : nums) {
            arr[num]++;
        }
        color_array(0, nums);
        color_array(1, nums);
        color_array(2, nums);

    }

};